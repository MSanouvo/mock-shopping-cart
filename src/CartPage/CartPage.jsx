import NavBar from "../components/Navbar/NavBar";
import CartItem from "../components/CartItems/CartItem";
import PropTypes from "prop-types";
import { useState } from "react";
import styles from "./CartPage.module.css";
import Button from "../components/button/button";

CartPage.propTypes = {
  cart: PropTypes.array,
  total: PropTypes.number
}

CartPage.defaultProps = {
  cart: [],
};

function CartPage(props) {
  const [cart, setCart] = useState(props.cart);

  const deleteItem = (item) => {
    const newCart = [...cart];
    for (let i = 0; i < newCart.length; i++) {
      if (newCart[i].name === item) {
        newCart.splice(i, 1);
        setCart(newCart);
      }
    }
    console.log(cart);
  };
  return (
    <div className={styles.container}>
      {/* <NavBar total={props.total} /> */}
      <main className={styles.cart}>
        <header className={styles.heading}>Shopping Cart</header>
        {cart.length != 0 ? (
          cart.map((item, index) => {
            return (
              <div key={index} className={styles.cartItems}>
                <CartItem
                  key={index}
                  name={item.name}
                  count={item.count}
                  src={item.src}
                  deleteItem={deleteItem}
                />
              </div>
            );
          })
        ) : (
          <p className={styles.message}>Add items to cart</p>
        )}
      </main>
      <div className={styles.button}>
        {cart.length != 0 && <Button name={"Complete Transaction"} />}
      </div>
    </div>
  );
}

export default CartPage;
