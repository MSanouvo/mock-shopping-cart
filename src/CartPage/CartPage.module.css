.container{
    /* border: 1px solid yellow; */
    min-width: 100vw;
}

.heading{
    font-weight: bold;
    font-size: 2.5rem;
    padding: 20px;
    color: black;
}

.cart{
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: start;
    gap: 20px;
    padding: 10px 20px;
    color: black;
    /* outline: 1px solid black; */
}

.cartItems{
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    gap: 30px;
    box-shadow: 0px 0px 5px black;
}

.button{
    display: flex;
    /* border: 1px solid green; */
    padding: 30px;
    width: 95%;
    justify-content: end;
}

.message{
    width: 100%;
    text-align: center;
}