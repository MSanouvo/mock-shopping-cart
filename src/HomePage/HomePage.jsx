import Banner from "../assets/clothes-line.jpg";
import styles from "./HomePage.module.css";

function HomePage() {
  return (
    <div className={styles.container}>
      <div className={styles.content}>
        <img className={styles.banner} src={Banner} alt="store banner" />
        <main className={styles.message}>
          <p>Welcome to my mock digital storefront!</p>
          <p className={styles.text}>
            Here you can shop for items by adding them to your cart and
            completing your transaction. You can begin by pressing on shop in
            the top right corner to see the store page. Once items have been
            added to the cart you can click on the cart icon to see them
            displayed there. If you change your mind about an item you can
            always delete it from the cart befor completing your transaction.
          </p>
        </main>
      </div>
    </div>
  );
}

export default HomePage;
