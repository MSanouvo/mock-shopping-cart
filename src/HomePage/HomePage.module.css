.banner {
  height: 50vh;
}

.container {
  /* border: 1px solid yellow; */
  min-width: 100vw;
}

.content {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 10px;
  padding: 30px 50px;
}

.message {
  font-size: 2rem;
  color: black;
  text-align: center;
}

.text {
  font-size: 1.2rem;
  width: 70vw;
}
