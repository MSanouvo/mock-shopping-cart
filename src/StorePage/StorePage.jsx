import StoreItem from "../components/StoreItems/StoreItem";
import PropTypes from "prop-types";
import { useState, useEffect } from "react";
import styles from "./StorePage.module.css";

StorePage.propTypes = {
  callItem: PropTypes.func
}

function StorePage(props) {
  const [inventory, setInventory] = useState(null);

  useEffect(() => {
    fetch("https://fakestoreapi.com/products/")
      .then((res) => res.json())
      .then((res) => setInventory(res.splice(0, 12)));
  }, []);

  return (
    <div className={styles.container}>
      <main className={styles.store}>
        <header className={styles.heading}>Store</header>
        <div className={styles.items}>
          {inventory != null ? (
            inventory.map((item, index) => {
              return (
                <div key={index}>
                  <StoreItem
                    key={index}
                    name={item.title}
                    count={1}
                    callItem={props.callItem}
                    src={item.image}
                    alt={item.title}
                  />
                </div>
              );
            })
          ) : (
            <p className={styles.message}>Loading items...</p>
          )}
        </div>
      </main>
    </div>
  );
}

export default StorePage;
