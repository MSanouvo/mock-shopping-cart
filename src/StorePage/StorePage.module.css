.container{
    /* border: 1px solid yellow; */
    display: flex;
    flex-direction: column;
    height: 100vh;
    min-width: 100vw;
    overflow-y: scroll;
}

.heading{
    font-weight: bold;
    font-size: 2.5rem;
    padding: 20px;
    color: black;
}

.store{
    display: flex;
    flex-direction: column;
    padding: 10px 20px;
}

.items{
    /* outline: 1px solid purple; */
    display: grid;
    grid-template-rows: 100px, 1fr;
    grid-template-columns: repeat(3, 30%);
    justify-content: center;
    height: 500px;
    gap: 30px;
    grid-row: 2;
    padding: 10px;
}

.message{
    width: 100%;
    text-align: center;
}