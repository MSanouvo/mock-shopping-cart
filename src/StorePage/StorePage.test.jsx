import { describe, it, expect, vi } from "vitest";
import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Route, Routes } from "react-router-dom";
import { MemoryRouter } from "react-router-dom";
import StorePage from "./StorePage";
import HomePage from "../HomePage/HomePage";
import CartPage from "../CartPage/CartPage";

describe("Home page", () => {
  it("renders the Home Page", () => {
    const { container } = render(
      <MemoryRouter>
        <StorePage />
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the heading components", () => {
    render(
      <MemoryRouter>
        <StorePage />
      </MemoryRouter>,
    );
    expect(screen.getByRole("heading", { name: "Home" })).toBeInTheDocument();
    expect(screen.getByRole("heading", { name: "Shop" })).toBeInTheDocument();
    expect(screen.getByAltText("cart icon")).toBeInTheDocument();
    expect(screen.getByText("Store")).toBeInTheDocument();
  });

  it("renders all the item card components", async () => {
    render(
      <MemoryRouter>
        <StorePage />
      </MemoryRouter>,
    );
    //No API Data
    expect(screen.getByText("Loading items...")).toBeInTheDocument();
    await waitForElementToBeRemoved(screen.getByText("Loading items..."));

    //Expected API Data
    expect(
      screen.getByRole("heading", {
        name: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
      }),
    ).toBeInTheDocument();
    expect(
      screen.getByAltText(
        "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
      ),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", {
        name: "Mens Casual Premium Slim Fit T-Shirts",
      }),
    ).toBeInTheDocument();
    expect(
      screen.getByAltText("Mens Casual Premium Slim Fit T-Shirts"),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "Mens Cotton Jacket" }),
    ).toBeInTheDocument();
    expect(screen.getByAltText("Mens Cotton Jacket")).toBeInTheDocument();

    expect(screen.getAllByRole("spinbutton").length).toEqual(3);
    expect(
      screen.getAllByRole("button", { name: "Add to Cart" }).length,
    ).toEqual(3);
  });

  it("navigates to home page", async () => {
    render(
      <MemoryRouter initialEntries={["/store"]}>
        <Routes>
          <Route path="/store" element={<StorePage />} />
          <Route path="/" element={<HomePage />} />
        </Routes>
      </MemoryRouter>,
    );

    const user = userEvent.setup();
    const homeLink = screen.getByRole("link", { name: "Home" });

    await user.click(homeLink);
    expect(
      screen.getByText("Welcome to my mock digital storefront!"),
    ).toBeInTheDocument();
  });

  it("navigates to cart page", async () => {
    render(
      <MemoryRouter initialEntries={["/store"]}>
        <Routes>
          <Route path="/store" element={<StorePage />} />
          <Route path="/cart" element={<CartPage cart={[]} />} />
        </Routes>
      </MemoryRouter>,
    );

    const user = userEvent.setup();
    const link = screen.getByAltText("cart icon");

    await user.click(link);
    expect(screen.getByText("Shopping Cart")).toBeInTheDocument();
  });

  it("has working add to cart button", async () => {
    const mock = vi.fn();
    const user = userEvent.setup();
    render(
      <MemoryRouter>
        <StorePage callItem={mock} />
      </MemoryRouter>,
    );

    expect(screen.getByText("Loading items...")).toBeInTheDocument();
    await waitForElementToBeRemoved(screen.getByText("Loading items..."));

    const button = screen.getAllByRole("button", { name: "Add to Cart" });
    await user.click(button[0]);
    expect(mock).toHaveBeenCalledWith({
      count: 1,
      name: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
      src: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
    });
  });
});
