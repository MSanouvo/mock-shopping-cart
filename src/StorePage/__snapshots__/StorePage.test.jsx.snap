// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home page > renders the Home Page 1`] = `
<div>
  <div
    class="_container_4b21f0"
  >
    <main
      class="_store_4b21f0"
    >
      <header
        class="_heading_4b21f0"
      >
        Store
      </header>
      <div
        class="_items_4b21f0"
      >
        <p
          class="_message_4b21f0"
        >
          Loading items...
        </p>
      </div>
    </main>
  </div>
</div>
`;
